Fixes #24408 - Added module-stream command

johnpmitsch opened the pull request:
Adding module-stream command.

**Relies on https://github.com/Katello/katello/pull/7638 and https://github.com/Katello/katello/pull/7688**

Here are some examples of the command:

Full list
```
[vagrant@centos7-hammer-devel hammer-cli-katello]$ hammer module-stream list
---|--------------------|-------------|--------------------------------------|----------------|--------------|---------
ID | MODULE STREAM NAME | STREAM      | UUID                                 | VERSION        | ARCHITECTURE | CONTEXT 
---|--------------------|-------------|--------------------------------------|----------------|--------------|---------
39 | ant                | 1.10        | db1d7b30-3db3-49c9-a141-1276dc0f8bb3 | 20180629154141 | x86_64       | 819b5873
31 | avocado            | latest      | 9e6a0c79-f838-40cf-b6c0-729f4ce23669 | 20180726175620 | x86_64       | 6c81f848
18 | avocado            | stable      | 4766ebb8-d340-4226-b221-cc81ac2f94f3 | 20180724154248 | x86_64       | 6c81f848
43 | container-tools    | 2018.0      | fd1bf17e-c56a-45f2-bd7e-bb634d69087b | 20180529173716 | x86_64       | cb85db27
37 | container-tools    | 2017.0      | bfaff19d-9c49-4156-bf2b-7f2f4a1e81c9 | 20180409175543 | x86_64       | ad2887e2
44 | cri-o              | 2018.0      | fe83c399-5bea-446c-8607-6a94a6217eb5 | 20180529183952 | x86_64       | 3ff668f0
13 | cri-o              | 2017.0      | 2d0bbeac-38ca-4068-adf4-6110d84d4d54 | 20180409175959 | x86_64       | eaa0eea8
45 | django             | 1.6         | 5f9e38c4-6a14-40be-b554-ffcc6763235e | 20180307130104 | noarch       | c2c572ec
38 | django             | 1.6         | d3ac99de-7d39-4113-af8d-4f400be1f05b | 20180409143454 | x86_64       | 6c81f848
42 | docker             | 2017.0      | f468895a-2a0d-4087-8c83-9d0fb1fdf276 | 20180409180031 | x86_64       | eaa0eea8
4  | duck               | 0           | 68065fcb-0777-478c-b16b-5c49fa6e2a1e | 20180704244205 | noarch       | deadbeef
1  | duck               | 0           | 2c5ebdc1-1504-4089-a318-c83ace3acdde | 20180730233102 | noarch       | deadbeef
10 | gcsf               | master      | 188a8ac4-86a4-47ed-9e9f-5400659be7f8 | 20180804153258 | x86_64       | b33445dc
19 | gimp               | 2.10        | 5809df22-65ef-4d2e-b556-1c90d8c4613e | 20180705113542 | x86_64       | 6c81f848
8  | golang             | 1.10        | 0e6a82d7-85aa-47fc-b749-9e0bd6a5b777 | 20180529132243 | x86_64       | 6c81f848
28 | golang-ecosystem   | 2018.0      | 93d1e298-e128-4673-9ee7-1b12dddef9e8 | 20180529151120 | x86_64       | cbfb2833
15 | golang-ecosystem   | 2017.0      | 34db5154-b522-4e68-9640-3fce5eb8e036 | 20180724154421 | x86_64       | cbfb2833
32 | hub                | pre-release | 9e6bf519-c8c6-45ba-9f44-bed57e6e33d4 | 20180809051024 | x86_64       | 6c81f848
5  | kangaroo           | 0           | a8a58a9c-b966-4f89-aca8-c0545beba013 | 20180730223407 | noarch       | deadbeef
3  | kangaroo           | 0           | 4f26fdb9-a84e-4f9d-841d-38d4e5ae016e | 20180704111719 | noarch       | deadbeef
34 | lizardfs           | devel       | b0fe6c6c-c7db-4de4-b48a-4aa70d0319c8 | 20180810084334 | x86_64       | 6c81f848
41 | mariadb            | 10.1        | e6f12097-7d0b-4e92-8cef-2f03a38865df | 20180418185803 | x86_64       | 6c81f848
29 | maven              | master      | 9c9df528-7061-4bb5-9af2-20507bd30e4c | 20180406164508 | x86_64       | 89d012dc
26 | maven              | 3.5         | 8226891f-f475-41b0-b919-d444fe12e2d5 | 20180629153919 | x86_64       | 819b5873
27 | meson              | master      | 85935202-8a9d-498e-8d83-c0eb3cebe916 | 20180812151830 | x86_64       | 06d0a27d
20 | mongodb            | 3.6         | 5ca4f696-0668-415b-9185-08638a3a4445 | 20180601084133 | x86_64       | 6c81f848
11 | mongodb            | 3.4         | 2959044d-6ee4-441e-a70f-4acfca6b6b13 | 20180601083501 | x86_64       | 6c81f848
14 | mysql              | 5.7         | 3380de31-bd6c-40f6-9836-3a7f8f953456 | 20180507085551 | x86_64       | 6c81f848
40 | ninja              | master      | e5f07fb4-cc63-4539-ba4c-44e6e2b29ac9 | 20180812151633 | x86_64       | 6c81f848
48 | nodejs             | 9           | 7a33f852-e388-40fd-8dc7-8efe86dd348a | 20180308142225 | noarch       | c2c572ec
47 | nodejs             | 8           | 6dcec8f2-1ec6-4dcb-a7b6-770d1ee9e49f | 20180308143646 | noarch       | c2c572ec
46 | nodejs             | 6           | e2e3cfa4-a2c4-4efc-92de-cc9310853297 | 20180308155546 | noarch       | c2c572ec
25 | nodejs             | 10          | 82108f9e-6800-4e88-ac5f-8ba5d39192ad | 20180807152904 | x86_64       | 6c81f848
22 | nodejs             | 8           | 61a481c7-5c1c-4cff-b783-dcd48b435772 | 20180724192622 | x86_64       | 6c81f848
36 | perl               | 5.24        | b49a224c-46e8-416e-89cf-62898d4b59da | 20180418132302 | x86_64       | 5d9ae238
9  | perl               | 5.26        | 181cc6de-19b4-4b5e-bf5e-674ee2d61218 | 20180417112647 | x86_64       | ccbeb30a
23 | perl-bootstrap     | 5.26        | 67a492a3-13e0-4473-9da3-4514966f3c85 | 20180417064707 | x86_64       | 6c81f848
21 | perl-bootstrap     | 5.24        | 5f6858eb-6fbf-434b-8693-7156e44cce69 | 20180417065503 | x86_64       | 6c81f848
7  | pki                | 10.6        | 083029c5-faee-4652-9162-0eb4d386a83d | 20180713022640 | x86_64       | 6c81f848
17 | postgresql         | 9.6         | 46d211d8-0786-4214-9950-bfb47b41d23e | 20180429200004 | x86_64       | 6c81f848
49 | reviewboard        | 2.5         | b9b5eb6a-ed15-410b-874f-80af68ad6e7e | 20180206144254 | noarch       | e0c83381
35 | reviewboard        | 3.0         | b4480c66-be88-45a8-90c0-18538bd7732b | 20180706125335 | x86_64       | 083bce86
24 | reviewboard        | 2.5         | 794da5e9-9248-415c-9f7e-bc5982eb5162 | 20180409150256 | x86_64       | 88f8ec0d
30 | ripgrep            | master      | 9d88cc75-da1b-4303-ab4c-d8fc45b4855c | 20180804140932 | x86_64       | b33445dc
33 | scala              | 2.10        | aa15541c-09f6-41fa-a9b0-5f568618ee15 | 20180702155503 | x86_64       | 819b5873
12 | stratis            | master      | 2b78fd0c-54e6-4d20-a0ea-d03860238fb6 | 20180726133047 | x86_64       | b33445dc
16 | testmodule         | master      | 3b9e659d-2fc0-4c27-baed-ec91f2f5ee0a | 20180405123256 | x86_64       | c2c572ec
6  | walrus             | 0.71        | b29c994c-bfc3-4887-a957-837b20b6c82a | 20180707144203 | x86_64       | c0ffee42
2  | walrus             | 5.21        | 31c7ab1e-0999-4dee-b1b2-6716b6adbfa2 | 20180704144203 | x86_64       | deadbeef
---|--------------------|-------------|--------------------------------------|----------------|--------------|---------
```

Full list for a repo
```
[vagrant@centos7-hammer-devel hammer-cli-katello]$ hammer module-stream list --repository-id 1
---|--------------------|--------|--------------------------------------|----------------|--------------|---------
ID | MODULE STREAM NAME | STREAM | UUID                                 | VERSION        | ARCHITECTURE | CONTEXT 
---|--------------------|--------|--------------------------------------|----------------|--------------|---------
4  | duck               | 0      | 68065fcb-0777-478c-b16b-5c49fa6e2a1e | 20180704244205 | noarch       | deadbeef
1  | duck               | 0      | 2c5ebdc1-1504-4089-a318-c83ace3acdde | 20180730233102 | noarch       | deadbeef
5  | kangaroo           | 0      | a8a58a9c-b966-4f89-aca8-c0545beba013 | 20180730223407 | noarch       | deadbeef
3  | kangaroo           | 0      | 4f26fdb9-a84e-4f9d-841d-38d4e5ae016e | 20180704111719 | noarch       | deadbeef
6  | walrus             | 0.71   | b29c994c-bfc3-4887-a957-837b20b6c82a | 20180707144203 | x86_64       | c0ffee42
2  | walrus             | 5.21   | 31c7ab1e-0999-4dee-b1b2-6716b6adbfa2 | 20180704144203 | x86_64       | deadbeef
---|--------------------|--------|--------------------------------------|----------------|--------------|---------
```

Info for a module stream
```
[vagrant@centos7-hammer-devel hammer-cli-katello]$ hammer module-stream info --id 28
ID:                 28
Module Stream Name: golang-ecosystem
Stream:             2018.0
UUID:               93d1e298-e128-4673-9ee7-1b12dddef9e8
Version:            20180529151120
Architecture:       x86_64
Context:            cbfb2833
Repositories:       
 1) ID:   9
    Name: rawhide modularity repo
Artifacts:          
 1) ID:   905
    Name: golang-github-russross-blackfriday-unit-test-0:2.0.0-1.20180215git55d61fa.module_1664+8cf66f9c.x86_64
 2) ID:   904
    Name: golang-github-russross-blackfriday-devel-0:2.0.0-1.20180215git55d61fa.module_1664+8cf66f9c.noarch
 3) ID:   903
    Name: golang-github-cpuguy83-go-md2man-devel-0:1.0.7-5.20180307git1d903dc.module_1664+8cf66f9c.noarch
 4) ID:   902
    Name: golang-github-cpuguy83-go-md2man-0:1.0.7-5.20180307git1d903dc.module_1664+8cf66f9c.x86_64
 5) ID:   901
    Name: go-srpm-macros-0:2-17.module_1812+d7358d9b.noarch
 6) ID:   900
    Name: go-compilers-golang-compiler-0:1-30.module_1812+d7358d9b.x86_64
Profiles:           
 1) ID:   43
    Name: default
    RPMs: 
     1) ID:   60
        Name: go-compilers
     2) ID:   61
        Name: go-srpm-macros
     3) ID:   62
        Name: golang-github-cpuguy83-go-md2man
     4) ID:   63
        Name: golang-github-russross-blackfriday


```

johnpmitsch commented:
tests are currently broken due to unrelated updates, these are being worked on. This is still reviewable.

johnpmitsch commented:
@akofink I see coveralls is decreased, i'm not sure what else I should be testing, but open to suggestions :)

akofink commented:
I don't care about 0.2% :)

akofink commented on lib/hammer_cli_katello/module_stream.rb:
```diff
@@ -0,0 +1,55 @@
+    resource :module_streams
+
+    class ListCommand < HammerCLIKatello::ListCommand
+      output do
```
`desc 'Whatever you want here'`

```
$ hammer module-stream -h
Usage:
    hammer module-stream [OPTIONS] SUBCOMMAND [ARG] ...

Parameters:
 SUBCOMMAND                    Subcommand
 [ARG] ...                     Subcommand arguments

Subcommands:
 info                          Show a module stream
 list                          List module_streams # I don't like this one

Options:
 -h, --help                    Print help
```

johnpmitsch commented:
FYI for reviewers - Currently CVV and CV flags are broken, this is true for the API parameters themselves in Katello. This is being added in separate changes, but we didn't want to let that block this work.

akofink commented on lib/hammer_cli_katello/module_stream.rb:
```diff
@@ -0,0 +1,55 @@
+        field :context, _("Context")
+      end
+
+      build_options
```
```sh
$ hammer module-stream list --repository modular --product-id 253
Error: Unrecognised option '--product-id'.
$ hammer module-stream list --repository modular
Error: Found more than one repository.
```

Note, this works fine if you don't have two products with the same repo name, but resolving repo name should always require product ID.

johnpmitsch replied:
would this have to happen on the API side of things? Currently we don't take product id for module streams info endpoint

![here](http://www.clipular.com/c/4879502698872832.png?k=MEC5GUshbc_hBk7je_XdgXOsXeg)

akofink replied:
No, this should be able to be enabled with the option builder. Something like:

```rb
build_options do |o|
  o.expand.including(:repositories)
  o.expand.including(:products) # or maybe this, I'm not sure
end
```

johnpmitsch replied:
That works for adding the option, but I don't think that fixes the original issue of having two repos in different products with the same name as it doesn't require the `--product-id` params when using `--repository`. 

This also creates (imo) some confusion for the product arguments. I'm seeing this in other repository content types as well so it may be a separate issue to address if I'm not misunderstanding something:

You can't use product-id by itself, but it is a specified option in the help menu. Maybe you can use it for a product with one repository, but that doesn't seem too useful.
```
[vagrant@hammer-devel2 ~]$ hammer package list --product-id 1
Error: Found more than one repository.
[vagrant@hammer-devel2 ~]$ hammer erratum list --product-id 1
Error: Found more than one repository.
``` 
You can query the repo with or without the product id, nothing is enforced.
```
[vagrant@hammer-devel2 ~]$ hammer erratum list --product-id 1 --repository zoo_with_modules
---|----------------|-------------|----------------------
ID | ERRATA ID      | TYPE        | TITLE                
---|----------------|-------------|----------------------
1  | RHEA-2012:0059 | enhancement | Duck_Kangaroo_Erratum
5  | RHEA-2012:0058 | enhancement | Gorilla_Erratum      
4  | RHEA-2012:0056 | security    | Bird_Erratum         
3  | RHEA-2012:0057 | security    | Bear_ErratumPARTHA   
2  | RHEA-2012:0055 | security    | Sea_Erratum          
---|----------------|-------------|----------------------
[vagrant@hammer-devel2 ~]$ hammer erratum list --repository zoo_with_modules
---|----------------|-------------|----------------------
ID | ERRATA ID      | TYPE        | TITLE                
---|----------------|-------------|----------------------
1  | RHEA-2012:0059 | enhancement | Duck_Kangaroo_Erratum
5  | RHEA-2012:0058 | enhancement | Gorilla_Erratum      
4  | RHEA-2012:0056 | security    | Bird_Erratum         
3  | RHEA-2012:0057 | security    | Bear_ErratumPARTHA   
2  | RHEA-2012:0055 | security    | Sea_Erratum          
---|----------------|-------------|----------------------
```

Maybe I'm misunderstanding how this is supposed to function, but product-id doesn't seem to be required when adding it to the build options. I'll keep it there for now since that seems to be the pattern for repository content classes, and will give the option to specify a product when there are repos of the same name, but it doesn't seem enforced.

akofink replied:
Ah, yes. The second piece is to add option validation. :) Sorry I didn't mention that before. [Here's an example](https://github.com/Katello/hammer-cli-katello/blob/49b48b3c1d431da10c254e3c59e62552cb909122/lib/hammer_cli_katello/content_view.rb#L112). Does that address your concerns?

johnpmitsch replied:
@akofink thanks, that does help. I'll add something like that. I'll probably abstract it out and include it in the other content type classes (errata, package group, etc...) as well since its applicable there.

akofink commented on test/functional/module_stream/list_test.rb:
```diff
@@ -0,0 +1,33 @@
+
+module HammerCLIKatello
+  describe ModuleStreamCommand::ListCommand do
+    it 'allows minimal options' do
```
```rb
it 'resolves repository ID from repo name and product ID'
it 'resolves repository ID from repo name, product name, and org ID'
it 'supports --search'
it 'allows --content-view-version-id'
```

johnpmitsch replied:
+1 to all of the above except search isn't implemented yet

akofink replied:
If search isn't implemented, and you aren't planning to implement it, we should remove that option from the subcommand.

akofink replied:
yeah, I'm fine with it. I really want to avoid releasing a version of hammer with these 'broken' options, so as long as all those other PRs make it into the release, we're fine.

akofink commented on test/functional/module_stream/list_test.rb:
```diff
@@ -0,0 +1,33 @@
+      run_cmd(%w(module-stream list --repository-id 1))
+    end
+
+    it 'can be provided by repository name' do
```
```rb
it 'requires product ID when repository name is provided'
```

akofink commented on lib/hammer_cli_katello/module_stream.rb:
```diff
@@ -0,0 +1,55 @@
+        field :context, _("Context")
+      end
+
+      build_options
```
```sh
$ hammer module-stream list --content-view-version-id 1
undefined method `module_streams' for #<Katello::ContentViewVersion:0x00007ff1a3cf1060>
Did you mean?  model_name
```

Make sure to check `hammer module-stream [subcommand] -h` and test all those options.

akofink commented:
You can ignore my CV/CVV comments :)

johnpmitsch commented:
@akofink I added requiring a product option anytime '--repository' is specified with a repository name to any content unit command. Please double check  I covered all of them. I've updated any existing tests to reflect this change. This is all in a separate commit so its easier to review :eyes:

johnpmitsch commented:
@akofink  I looked into this error that we discussed yesterday:
```
[vagrant@hammer-devel2 ~]$ hammer package-group list --product-id 1
Error: Found more than one repository.
```

I don't see `product_id` as a parameter for `package_groups` or any other repo content unit in the API. I think hammer just puts `--product-id` to specify the product when specifying the repository name. The error message could perhaps be clearer, but I don't think further action is needed there. Curious to hear what you think

akofink commented:
Yes, I think the option builder is adding product-id for resolving repository name. If we want to use it as an option (which, this issue doesn't need to address), I think we might have to explicitly add the option, [like this](https://github.com/Katello/hammer-cli-katello/blob/49b48b3c1d431da10c254e3c59e62552cb909122/lib/hammer_cli_katello/content_view.rb#L104), or use a [custom option builder](https://github.com/theforeman/hammer-cli/blob/master/doc/option_builders.md).

johnpmitsch commented:
@akofink naming things is indeed hard XD I updated to the long and verbose, but very clear `validate_repo_name_requires_product_options`

akofink commented on lib/hammer_cli_katello/file.rb:
```diff
@@ -3,6 +3,10 @@ class FileCommand < HammerCLIKatello::Command
 
     class ListCommand < HammerCLIKatello::ListCommand
+      extend RepositoryScopedToProduct
+
```
Use `desc` to set the description:

```sh
$ hammer -unonadmin -pchangeme module-stream -h
Usage:
    hammer module-stream [OPTIONS] SUBCOMMAND [ARG] ...

Parameters:
 SUBCOMMAND                    Subcommand
 [ARG] ...                     Subcommand arguments

Subcommands:
 info                          Show a module stream
 list                          List module_streams # this one isn't quite right

Options:
 -h, --help                    Print help
```

akofink commented on test/functional/module_stream/list_test.rb:
```diff
@@ -0,0 +1,50 @@
+    end
+
+    it 'can be provided by repository ID' do
+      run_cmd(%w(module-stream list --repository-id 1))
```
missing an expectation

akofink commented on test/functional/module_stream/list_test.rb:
```diff
@@ -0,0 +1,50 @@
+    end
+
+    it 'product params needed with repository name' do
+      cmd = run_cmd(%w(module-stream list --repository Repo))
```
missing expectation

johnpmitsch replied:
What is the expectation for the API call here? I'm not seeing an API call being made

```
[vagrant@hammer-devel2 ~]$ hammer -d  module-stream list --repository zoo_with_modules
[ INFO 2018-09-04T14:56:57 Init] Initialization of Hammer CLI (0.14.pre.develop) has started...
[DEBUG 2018-09-04T14:56:57 Init] Running at ruby 2.4.3-p205
[ INFO 2018-09-04T14:56:57 Init] Configuration from the file /home/vagrant/.hammer/cli_config.yml has been loaded
[ INFO 2018-09-04T14:56:57 Init] Configuration from the file /home/vagrant/.hammer/cli.modules.d/csv.yml has been loaded
[ INFO 2018-09-04T14:56:57 Init] Configuration from the file /home/vagrant/.hammer/cli.modules.d/foreman.yml has been loaded
[ INFO 2018-09-04T14:56:57 Init] Configuration from the file /home/vagrant/.hammer/cli.modules.d/foreman_admin.yml has been loaded
[ INFO 2018-09-04T14:56:57 Init] Configuration from the file /home/vagrant/.hammer/cli.modules.d/foreman_admin_logging_core.yml has been loaded
[ INFO 2018-09-04T14:56:57 Init] Configuration from the file /home/vagrant/.hammer/cli.modules.d/foreman_admin_logging_katello.yml has been loaded
[ INFO 2018-09-04T14:56:57 Init] Configuration from the file /home/vagrant/.hammer/cli.modules.d/katello.yml has been loaded
[ INFO 2018-09-04T14:56:57 SSLoptions] Matching CA cert was found in local CA store /home/vagrant/.hammer/certs/coffee.jomitsch.example.com_443.pem
[DEBUG 2018-09-04T14:56:57 SSLoptions] SSL options: {
     :verify_ssl => true,
    :ssl_ca_file => "/home/vagrant/.hammer/certs/coffee.jomitsch.example.com_443.pem"
}
[DEBUG 2018-09-04T14:56:57 API] Global headers: {
        :content_type => "application/json",
              :accept => "application/json;version=2",
    "Accept-Language" => "en"
}
[DEBUG 2018-09-04T14:56:57 API] Follow redirects: never
[DEBUG 2018-09-04T14:56:57 Connection] Registered: foreman
[ INFO 2018-09-04T14:56:57 Modules] Extension module hammer_cli_csv (2.3.1) loaded.
[ INFO 2018-09-04T14:56:57 Modules] Extension module hammer_cli_foreman (0.14.pre.develop) loaded.
[ INFO 2018-09-04T14:56:57 Modules] Extension module hammer_cli_foreman_admin (0.0.8) loaded.
[ INFO 2018-09-04T14:56:57 HammerCLI::MainCommand] subcommand organization (HammerCLIForeman::Organization) was removed.
[ INFO 2018-09-04T14:56:57 HammerCLI::MainCommand] subcommand organization (HammerCLIKatello::Organization) was created.
[ INFO 2018-09-04T14:56:57 Modules] Extension module hammer_cli_katello (0.14.0) loaded.
[DEBUG 2018-09-04T14:56:57 Init] Using locale 'en'
[DEBUG 2018-09-04T14:56:57 Init] 'mo' files for locale domain 'hammer-cli' loaded from '/home/vagrant/hammer-cli/locale'
[DEBUG 2018-09-04T14:56:57 Init] 'mo' files for locale domain 'hammer-cli-foreman' loaded from '/home/vagrant/hammer-cli-foreman/locale'
[DEBUG 2018-09-04T14:56:57 Init] 'mo' files for locale domain 'hammer-cli-csv' loaded from '/home/vagrant/hammer-cli-csv/locale'
[DEBUG 2018-09-04T14:56:57 Init] 'mo' files for locale domain 'hammer-cli-katello' loaded from '/home/vagrant/hammer-cli-katello/locale'
[ INFO 2018-09-04T14:56:57 HammerCLI::MainCommand] Called with options: {"option_debug"=>true}
[ INFO 2018-09-04T14:56:57 HammerCLIKatello::ModuleStreamCommand] Called with options: {}
[DEBUG 2018-09-04T14:56:57 Exception] Using exception handler HammerCLIKatello::ExceptionHandler#handle_usage_exception
[ERROR 2018-09-04T14:56:57 Exception] Error: At least one of options --product, --product-id is required.

See: 'hammer module-stream list --help'.
Error: At least one of options --product, --product-id is required.

See: 'hammer module-stream list --help'.
[ERROR 2018-09-04T14:56:57 Exception] 

```

akofink replied:
```rb
api_expects_no_call
```

akofink commented on test/functional/package_group/list_test.rb:
```diff
@@ -17,17 +17,21 @@ module HammerCLIKatello
-      end
-      ex.returns(index_response([{'id' => 1}]))
+    it 'needs product options with repository name' do
+      cmd = run_cmd(%w(package-group list --repository Repo))
```
missing expectation

akofink commented on test/functional/module_stream/info_test.rb:
```diff
@@ -0,0 +1,12 @@
+
+module HammerCLIKatello
+  describe ModuleStreamCommand::InfoCommand do
+    it 'allows ID' do
```
it 'resolves ID from name and product ID'
it 'resolves ID from name, product name, and repository ID'

johnpmitsch replied:
unfortunately, search for module streams isn't implemented, so I'm having trouble testing these options as they don't work quite right as is. We can delay this PR until search gets implemented, or add tests later.

akofink commented on lib/hammer_cli_katello/module_stream.rb:
```diff
@@ -0,0 +1,61 @@
+      end
+    end
+
+    class InfoCommand < HammerCLIKatello::InfoCommand
```
Missing product options for resolving repo ID from repo name.

```sh
$ hammer module-stream info -h
Usage:
    hammer module-stream info [OPTIONS]

Options:
 --id ID                       A module stream identifier
 --name NAME                   Name to search by
 --repository REPOSITORY_NAME  Repository name to search by
 --repository-id REPOSITORY_ID Repository ID
 -h, --help                    Print help

$ hammer module-stream info --name django --product custom --repository modular --organization-id 1
Error: Unrecognised option '--product'.

See: 'hammer module-stream info --help'.
```

Needs similar validation to list.

johnpmitsch replied:
added product options for info and a test as well

johnpmitsch commented:
@akofink updated according to your suggestions

akofink commented:
I don't understand why we can't fix and test this:

```
$ hammer module-stream info --name django --product custom --repository modular
Error: Could not find organization. Some search options were missing, please see --help.

$ hammer module-stream info --name django --product custom --repository modular --organization-id 1
Error: Unrecognised option '--organization-id'.

See: 'hammer module-stream info --help'.
```

johnpmitsch commented:
@akofink I think I missed the part about the organization id. I'll add that in. I still think there is an issue testing this, but I switched contexts, let me comment back when I can tell you more of the specifics

johnpmitsch commented:
@akofink updated to fix the info command's search added more tests

akofink commented on test/functional/module_stream/info_test.rb:
```diff
@@ -0,0 +1,58 @@
+      api_expects(:module_streams, :show)
+        .with_params(repository_id: 1, id: 1)
+
+      run_cmd(%w(module-stream info --name walrus --repository zoo
```
We're not expecting this functionality to actually work because search fails on module_streams, right?

johnpmitsch replied:
I should have specified that [this PR](https://github.com/Katello/katello/pull/7688) implements search and I was testing with that. Its currently under review

johnpmitsch commented:
@akofink is that an issue with this PR or just the nature of searching on a non-unique field? The only way to find a unique module stream is through VSCAN (version, stream, context, arch, and name)

akofink commented:
The index request should send module stream name and the repository ID at least. This is how every ID resolution in hammer currently [should] work.

johnpmitsch commented:
@akofink this is what I was asking yesterday:
```
 jomitsch | akofink: trying to figure something else out with hammer, when I submit with repository-id, I don'
          | see it actually used, it just searches by name in module_streams                                  
 jomitsch | https://paste.fedoraproject.org/paste/Zdu9UlWLcO9kvzyVnYhzTA                                      
 jomitsch | seems like the repo isn't used                                                                    
 jomitsch | I would expect it to pass 'repository-id' param or use the repository/:id/module_streams endpoint 
```
I don't know the correct way to go about handling this, it seems like it could be a bug?

akofink commented:
@johnpmitsch Yes, this is a pretty bad bug (for most content types, file_units are doing it mostly right). The following change at least passes repo ID to the search for module streams:

```diff
diff --git a/lib/hammer_cli_katello/foreman_search_options_creators.rb b/lib/hammer_cli_katello/foreman_search_options_creators.rb
index fe78492..ecf1df6 100644
--- a/lib/hammer_cli_katello/foreman_search_options_creators.rb
+++ b/lib/hammer_cli_katello/foreman_search_options_creators.rb
@@ -79,9 +79,5 @@ module HammerCLIKatello
     def create_image_search_options(options, mode = nil)
       create_search_options_without_katello_api(options, api.resource(:images), mode)
     end
-
-    def create_module_streams_search_options(options, mode = nil)
-      create_search_options_without_katello_api(options, api.resource(:module_streams), mode)
-    end
   end
 end
diff --git a/lib/hammer_cli_katello/id_resolver.rb b/lib/hammer_cli_katello/id_resolver.rb
index 316faf0..1e13914 100644
--- a/lib/hammer_cli_katello/id_resolver.rb
+++ b/lib/hammer_cli_katello/id_resolver.rb
@@ -10,6 +10,8 @@ module HammerCLIKatello
       :file_unit =>             [s_name(_("File name to search by")),
                                  s("content_view_version_id", _("Content View Version ID")),
                                  s("repository_id", _("Repository ID"))],
+      :module_stream =>         [s_name(_("Module stream name to search by")),
+                                 s("repository_id", _("Repository ID"))],
       :gpg =>                   [s_name(_("Gpg key name to search by"))],
       :host_collection =>       [s_name(_("Host collection name to search by"))],
       :lifecycle_environment => [s_name(_("Lifecycle environment name to search by"))],
diff --git a/lib/hammer_cli_katello/search_options_creators.rb b/lib/hammer_cli_katello/search_options_creators.rb
index fb44d6e..16f4a81 100644
--- a/lib/hammer_cli_katello/search_options_creators.rb
+++ b/lib/hammer_cli_katello/search_options_creators.rb
@@ -4,6 +4,11 @@ module HammerCLIKatello
   module SearchOptionsCreators
     include HammerCLIKatello::ForemanSearchOptionsCreators

+    def create_module_streams_search_options(options, mode = nil)
+      create_search_options_without_katello_api(options, api.resource(:module_streams), mode)
+        .merge(create_search_options(options, api.resource(:module_streams), mode))
+    end
+
     def create_file_units_search_options(options, mode = nil)
       create_search_options_without_katello_api(options, api.resource(:file_units), mode)
         .merge(create_search_options(options, api.resource(:file_units), mode))
```

```sh
[vagrant@centos7-hammer-devel hammer-cli-katello]$ hammer -rd module-stream info --name django --repository-id 14
[ INFO 2018-09-12T16:25:59 Init] Initialization of Hammer CLI (0.15.pre.develop) has started...
[DEBUG 2018-09-12T16:25:59 Init] Running at ruby 2.4.3-p205
[ INFO 2018-09-12T16:25:59 Init] Configuration from the file /home/vagrant/.hammer/cli_config.yml has been loaded
[ INFO 2018-09-12T16:25:59 Init] Configuration from the file /home/vagrant/.hammer/cli.modules.d/csv.yml has been loaded
[ INFO 2018-09-12T16:25:59 Init] Configuration from the file /home/vagrant/.hammer/cli.modules.d/foreman.yml has been loaded
[ INFO 2018-09-12T16:25:59 Init] Configuration from the file /home/vagrant/.hammer/cli.modules.d/foreman_admin.yml has been loaded
[ INFO 2018-09-12T16:25:59 Init] Configuration from the file /home/vagrant/.hammer/cli.modules.d/foreman_admin_logging_core.yml has been loaded
[ INFO 2018-09-12T16:25:59 Init] Configuration from the file /home/vagrant/.hammer/cli.modules.d/foreman_admin_logging_katello.yml has been loaded
[ INFO 2018-09-12T16:25:59 Init] Configuration from the file /home/vagrant/.hammer/cli.modules.d/katello.yml has been loaded
[ INFO 2018-09-12T16:26:00 SSLoptions] Matching CA cert was found in local CA store /home/vagrant/.hammer/certs/centos7-devel.virbr0.akofink-desktop_443.pem
[DEBUG 2018-09-12T16:26:00 SSLoptions] SSL options: {
     :verify_ssl => true,
    :ssl_ca_file => "/home/vagrant/.hammer/certs/centos7-devel.virbr0.akofink-desktop_443.pem"
}
[DEBUG 2018-09-12T16:26:00 API] Global headers: {
        :content_type => "application/json",
              :accept => "application/json;version=2",
    "Accept-Language" => "en"
}
[DEBUG 2018-09-12T16:26:00 API] Follow redirects: never
[DEBUG 2018-09-12T16:26:00 API] Apipie cache was cleared
[DEBUG 2018-09-12T16:26:00 Connection] Registered: foreman
[ INFO 2018-09-12T16:26:00 API] Server: https://centos7-devel.virbr0.akofink-desktop:443
[ INFO 2018-09-12T16:26:00 API] GET /apidoc/v2.en.json
[DEBUG 2018-09-12T16:26:00 API] Params: {}
[DEBUG 2018-09-12T16:26:00 API] Headers: {
    :accept => "application/json",
    :params => {}
}
[DEBUG 2018-09-12T16:26:01 API] Response: Received OK
[DEBUG 2018-09-12T16:26:01 API] Response headers: {
                                 :date => "Wed, 12 Sep 2018 16:26:00 GMT",
                               :server => "Apache/2.4.6 (CentOS)",
                         :content_type => "text/html",
                  :content_disposition => "inline; filename=\"v2.en.json\"",
            :content_transfer_encoding => "binary",
                        :cache_control => "private",
                      :apipie_checksum => "9058d31b5aa482eb27d6a3fa1e153362dc68199b",
                         :x_request_id => "5fc8ceb6-a6be-42a7-822d-18742183c11a",
                            :x_runtime => "1.617003",
              :content_security_policy => "default-src 'self'; child-src 'self'; connect-src 'self' ws: wss:; img-src 'self' data: *.gravatar.com; script-src 'unsafe-eval' 'unsafe-inline' 'self'; style-src 'unsafe-inline' 'self'",
            :strict_transport_security => "max-age=631139040; includeSubdomains",
               :x_content_type_options => "nosniff",
                   :x_download_options => "noopen",
                      :x_frame_options => "sameorigin",
    :x_permitted_cross_domain_policies => "none",
                     :x_xss_protection => "1; mode=block",
                                  :via => "1.1 centos7-devel.virbr0.akofink-desktop",
                                 :vary => "Accept-Encoding",
                     :content_encoding => "gzip",
                    :transfer_encoding => "chunked",
                     :content_language => "en"
}
[DEBUG 2018-09-12T16:26:01 API] New apidoc loaded from the server
[ INFO 2018-09-12T16:26:02 Modules] Extension module hammer_cli_csv (2.3.1) loaded.
[ INFO 2018-09-12T16:26:02 Modules] Extension module hammer_cli_foreman (0.15.pre.develop) loaded.
[ INFO 2018-09-12T16:26:02 Modules] Extension module hammer_cli_foreman_admin (0.0.8) loaded.
[ INFO 2018-09-12T16:26:02 HammerCLI::MainCommand] subcommand organization (HammerCLIForeman::Organization) was removed.
[ INFO 2018-09-12T16:26:02 HammerCLI::MainCommand] subcommand organization (HammerCLIKatello::Organization) was created.
[ INFO 2018-09-12T16:26:02 Modules] Extension module hammer_cli_katello (0.14.0) loaded.
[DEBUG 2018-09-12T16:26:02 Init] Using locale 'en'
[DEBUG 2018-09-12T16:26:02 Init] 'mo' files for locale domain 'hammer-cli' loaded from '/home/vagrant/hammer-cli/locale'
[DEBUG 2018-09-12T16:26:02 Init] 'mo' files for locale domain 'hammer-cli-foreman' loaded from '/home/vagrant/hammer-cli-foreman/locale'
[DEBUG 2018-09-12T16:26:02 Init] 'mo' files for locale domain 'hammer-cli-csv' loaded from '/home/vagrant/hammer-cli-csv/locale'
[DEBUG 2018-09-12T16:26:02 Init] 'mo' files for locale domain 'hammer-cli-katello' loaded from '/home/vagrant/hammer-cli-katello/locale'
[ INFO 2018-09-12T16:26:02 HammerCLI::MainCommand] Called with options: {"option_debug"=>true, "option_reload_cache"=>true}
[ INFO 2018-09-12T16:26:02 HammerCLIKatello::ModuleStreamCommand] Called with options: {}
[ INFO 2018-09-12T16:26:02 API] Server: https://centos7-devel.virbr0.akofink-desktop:443
[ INFO 2018-09-12T16:26:02 API] GET /katello/api/repositories/14/module_streams
[DEBUG 2018-09-12T16:26:02 API] Params: {
    :search => "name = \"django\"",
     "name" => "django"
}
[DEBUG 2018-09-12T16:26:02 API] Headers: {
    :params => {
        :search => "name = \"django\"",
         "name" => "django"
    }
}
[DEBUG 2018-09-12T16:26:02 API] Using authenticator: HammerCLIForeman::Api::InteractiveBasicAuth
[DEBUG 2018-09-12T16:26:11 API] Response: {
       "total" => 5,
    "subtotal" => 1,
        "page" => 1,
    "per_page" => 20,
       "error" => nil,
      "search" => "name = \"django\"",
        "sort" => {
           "by" => "name",
        "order" => "asc"
    },
     "results" => [
        [0] {
                      "id" => 3,
                    "name" => "django",
                    "uuid" => "c637a7ef-6d30-4a9b-aed5-e7fc073fd9ba",
                 "version" => "20180307130104",
                 "context" => "c2c572ec",
                  "stream" => "1.6",
                    "arch" => "noarch",
             "description" => "Django is a high-level Python Web framework that encourages rapid development and a clean, pragmatic design. It focuses on automating as much as possible and adhering to the DRY (Don't Repeat Yourself) principle.",
                 "summary" => "A high-level Python Web framework",
            "repositories" => [
                [0] {
                      "id" => 14,
                    "name" => "modular"
                }
            ]
        }
    ]
}
[DEBUG 2018-09-12T16:26:11 API] Response headers: {
                                 :date => "Wed, 12 Sep 2018 16:26:02 GMT",
                               :server => "Apache/2.4.6 (CentOS)",
                      :foreman_version => "1.20.0-develop",
                  :foreman_api_version => "2",
                         :content_type => "application/json; charset=utf-8",
                      :apipie_checksum => "9058d31b5aa482eb27d6a3fa1e153362dc68199b",
                                 :etag => "W/\"ea9151754ad28d5f7616e090996af636-gzip\"",
                        :cache_control => "max-age=0, private, must-revalidate",
                         :x_request_id => "629869ca-a843-4bb4-abc1-1aa91d5909f4",
                            :x_runtime => "8.859581",
              :content_security_policy => "default-src 'self'; child-src 'self'; connect-src 'self' ws: wss:; img-src 'self' data: *.gravatar.com; script-src 'unsafe-eval' 'unsafe-inline' 'self'; style-src 'unsafe-inline' 'self'",
            :strict_transport_security => "max-age=631139040; includeSubdomains",
               :x_content_type_options => "nosniff",
                   :x_download_options => "noopen",
                      :x_frame_options => "sameorigin",
    :x_permitted_cross_domain_policies => "none",
                     :x_xss_protection => "1; mode=block",
                           :set_cookie => [
        [0] "_session_id=59412d171f218e0d5f3c8037d558d7f7; path=/; HttpOnly; secure; SameSite=Lax"
    ],
                                  :via => "1.1 centos7-devel.virbr0.akofink-desktop",
                                 :vary => "Accept-Encoding",
                     :content_encoding => "gzip",
                       :content_length => "408"
}
[ INFO 2018-09-12T16:26:11 HammerCLIKatello::ModuleStreamCommand::InfoCommand] Called with options: {"option_name"=>"django", "option_repository_id"=>"14", "option_id"=>3}
[ INFO 2018-09-12T16:26:11 API] Server: https://centos7-devel.virbr0.akofink-desktop:443
[ INFO 2018-09-12T16:26:11 API] GET /katello/api/repositories/14/module_streams/3
[DEBUG 2018-09-12T16:26:11 API] Params: {}
[DEBUG 2018-09-12T16:26:11 API] Headers: {
    :params => {}
}
[DEBUG 2018-09-12T16:26:11 API] Using authenticator: HammerCLIForeman::Api::InteractiveBasicAuth
[DEBUG 2018-09-12T16:26:12 API] Response: {
              "id" => 3,
            "name" => "django",
            "uuid" => "c637a7ef-6d30-4a9b-aed5-e7fc073fd9ba",
         "version" => "20180307130104",
         "context" => "c2c572ec",
          "stream" => "1.6",
            "arch" => "noarch",
     "description" => "Django is a high-level Python Web framework that encourages rapid development and a clean, pragmatic design. It focuses on automating as much as possible and adhering to the DRY (Don't Repeat Yourself) principle.",
         "summary" => "A high-level Python Web framework",
    "repositories" => [
        [0] {
              "id" => 14,
            "name" => "modular"
        }
    ],
       "artifacts" => [
        [0] {
              "id" => 15,
            "name" => "python2-django-0:1.6.11.7-1.module_1560+089ce146.noarch"
        },
        [1] {
              "id" => 14,
            "name" => "python-django-bash-completion-0:1.6.11.7-1.module_1560+089ce146.noarch"
        }
    ],
        "profiles" => [
        [0] {
              "id" => 6,
            "name" => "default",
            "rpms" => [
                [0] {
                      "id" => 9,
                    "name" => "python2-django"
                }
            ]
        },
        [1] {
              "id" => 8,
            "name" => "python2_development",
            "rpms" => [
                [0] {
                      "id" => 11,
                    "name" => "python2-django"
                }
            ]
        }
    ]
}
[DEBUG 2018-09-12T16:26:12 API] Response headers: {
                                 :date => "Wed, 12 Sep 2018 16:26:11 GMT",
                               :server => "Apache/2.4.6 (CentOS)",
                      :foreman_version => "1.20.0-develop",
                  :foreman_api_version => "2",
                         :content_type => "application/json; charset=utf-8",
                      :apipie_checksum => "9058d31b5aa482eb27d6a3fa1e153362dc68199b",
                                 :etag => "W/\"462b0a18c6acd7a7ae97f9f9447ec513-gzip\"",
                        :cache_control => "max-age=0, private, must-revalidate",
                         :x_request_id => "8773abda-751d-4773-8eba-cb76e6d8af65",
                            :x_runtime => "1.294374",
              :content_security_policy => "default-src 'self'; child-src 'self'; connect-src 'self' ws: wss:; img-src 'self' data: *.gravatar.com; script-src 'unsafe-eval' 'unsafe-inline' 'self'; style-src 'unsafe-inline' 'self'",
            :strict_transport_security => "max-age=631139040; includeSubdomains",
               :x_content_type_options => "nosniff",
                   :x_download_options => "noopen",
                      :x_frame_options => "sameorigin",
    :x_permitted_cross_domain_policies => "none",
                     :x_xss_protection => "1; mode=block",
                           :set_cookie => [
        [0] "_session_id=5def3cfc1483fb223ac6f0e5bb5bd71f; path=/; HttpOnly; secure; SameSite=Lax"
    ],
                                  :via => "1.1 centos7-devel.virbr0.akofink-desktop",
                                 :vary => "Accept-Encoding",
                     :content_encoding => "gzip",
                       :content_length => "449"
}
ID:                 3
Module Stream Name: django
Stream:             1.6
UUID:               c637a7ef-6d30-4a9b-aed5-e7fc073fd9ba
Version:            20180307130104
Architecture:       noarch
Context:            c2c572ec
Repositories:
 1) ID:   14
    Name: modular
Artifacts:
 1) ID:   15
    Name: python2-django-0:1.6.11.7-1.module_1560+089ce146.noarch
 2) ID:   14
    Name: python-django-bash-completion-0:1.6.11.7-1.module_1560+089ce146.noarch
Profiles:
 1) ID:   6
    Name: default
    RPMs:
     1) ID:   9
        Name: python2-django
 2) ID:   8
    Name: python2_development
    RPMs:
     1) ID:   11
        Name: python2-django
```

I will report this issue for other content types, mentioning the need to add the search options for VSCAN as well (good catch!).

johnpmitsch commented:
@akofink updated!
